### PR TITLE
feat: add checksum verification for checkpoint

### PR DESCRIPTION
**Pull request by CrystalAnalyst**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

in file: src/mito2/src/manifest/storage.rs. 
[goal]: to add checksum for manifest, since there's 2 types manifest: delta and checkpoint, this one **add checksum for checkpoint**.
related issue: #2164 

Please explain IN DETAIL what the changes are in this PR and why they are needed:

- Summarize your change (**mandatory**)
  step1. using crc32fast to add checksum(after compress and encoding)in function: `save_checkpoint()`.
  step2. calculate the checksum and make comparison with stored_checksum in function: `load_checkpoint()`.
- How does this PR work? Need a brief introduction for the changed logic (optional)
- Describe clearly one logical change and avoid lazy messages (optional)
- Describe any limitations of the current code (optional)

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

**Comment by CrystalAnalyst** [comment]
@hygkui @killme2008 @sunng87 @lbt05 please take a look, thanks! any suggestion for improvement is welcomed.

**Comment by CrystalAnalyst** [comment]
seems there are some changes need to be made, I'll work on them in the next few days.

**Comment by CrystalAnalyst** [comment]
I‘ve done and solved the conflict.

**Comment by CrystalAnalyst** [comment]
> I‘ve done and solved the conflict.

@WenyXu , dear Weny, please take a look.

File: src/mito2/src/error.rs [hunk 1]
```diff
@@ -485,6 +484,12 @@ pub enum Error {
         flushed_entry_id: u64,
         unexpected_entry_id: u64,
     },
+    #[snafu(display(
+        "Verify checksum failed! calculated: {}, expected: {}",
+        calculated,
+        expect
+    ))]
+    VerifyChecksum { calculated: String, expect: String },
```
Comment (waynexia) [inline src/mito2/src/error.rs:492]: Nit: add empty line
```suggestion

    #[snafu(display(
        "Verify checksum failed! calculated: {}, expected: {}",
        calculated,
        expect
    ))]
    VerifyChecksum { calculated: String, expect: String },
```

File: src/mito2/src/manifest/tests/checkpoint.rs
Comment (waynexia) [inline src/mito2/src/manifest/tests/checkpoint.rs:1]: Can you add a test for corrupt data? Like change the content in checkpoint and check if it fails with checksum error

File: src/mito2/src/manifest/storage.rs [hunk 2]
```diff
@@ -367,6 +368,11 @@ impl ManifestObjectStore {
                 compress_type: self.compress_type,
                 path: &path,
             })?;
+
+        let mut hasher = Hasher::new();
+        hasher.update(&data);
+        let checksum = hasher.finalize();
```
Comment (waynexia) [inline src/mito2/src/manifest/storage.rs:374]: style: can you wrap this into a little util function?

File: src/mito2/src/manifest/storage.rs [hunk 3]
```diff
@@ -405,9 +411,23 @@ impl ManifestObjectStore {
...
+        };
+        let metadata = CheckpointMetadata::decode(&metadata_bytes)?;
+
+        let checkpoint_data = match self.object_store.read(&path).await {
+            Ok(checkpoint) => {
+                let mut hasher = Hasher::new();
+                hasher.update(&checkpoint);
+                let calculated_checksum = hasher.finalize();
+
+                if Some(format!("{}", calculated_checksum)) == metadata.checksum {
```
Comment (waynexia) [inline src/mito2/src/manifest/storage.rs:430]: What if `metadata.checksum` is `None`? We should not fail in this case

Comment (CrystalAnalyst) [reply src/mito2/src/manifest/storage.rs:430]: Thanks for your valuable suggestions!  I'll work on them.

File: src/mito2/src/error.rs (hunk 1 above)
Comment (CrystalAnalyst) [reply src/mito2/src/error.rs:492]: done

File: src/mito2/Cargo.toml [hunk 4]
```diff
@@ -32,6 +32,7 @@ common-runtime.workspace = true
 common-telemetry.workspace = true
 common-test-util = { workspace = true, optional = true }
 common-time.workspace = true
+crc32fast = "1.3.2"
```
Comment (evenyag) [inline src/mito2/Cargo.toml:35]:
```suggestion
crc32fast = "1"
```

File: src/mito2/src/error.rs [hunk 5]
```diff
@@ -533,6 +533,13 @@ pub enum Error {
         error: std::io::Error,
         location: Location,
     },
+
+    #[snafu(display(
+        "Verify checksum failed! calculated: {}, expected: {}",
+        calculated,
+        expect
+    ))]
+    VerifyChecksum { calculated: String, expect: String },
```
Comment (evenyag) [inline src/mito2/src/error.rs:542]: Consider adding a location.
```suggestion
    VerifyChecksum {
        calculated: String,
        expect: String,
        location: Location,
    },
```

File: src/mito2/src/manifest/storage.rs [hunk 6]
```diff
@@ -380,7 +392,7 @@ impl ManifestObjectStore {
         let checkpoint_metadata = CheckpointMetadata {
             size: bytes.len(),
             version,
-            checksum: None,
+            checksum: Some(format!("{}", checksum)),
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:395]: nitpick: I believe that we could use `to_string()`.
```suggestion
            checksum: Some(checksum.to_string()),
```

File: src/mito2/src/manifest/storage.rs [hunk 7]
```diff
@@ -405,9 +417,42 @@ impl ManifestObjectStore {
...
-            match self.object_store.read(&path).await {
-                Ok(checkpoint) => {
+
+        // get the stored_checksum, which is in the checkpoint metadata.
+        let metadata_path = self.last_checkpoint_path();
+        let metadata_bytes = match self.object_store.read(&metadata_path).await {
+            Ok(metadata_bytes) => metadata_bytes,
+            Err(e) if e.kind() == ErrorKind::NotFound => return Ok(None), // No metadata found
+            Err(e) => return Err(e).context(OpenDalSnafu),                // Error reading metadata
+        };
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:427]: Consider providing a helper method (e.g. `read_bytes()`) to read bytes from the object store and return `Ok(None)` if the object is not found.
```suggestion
        let Some(metadata_bytes) = self.read_bytes(&metadata_path).await? else {
            return Ok(None); // No metadata found
        };
```

We could use this method in other places below.

File: src/mito2/src/manifest/storage.rs [hunk 8]
```diff
@@ -405,9 +417,42 @@ impl ManifestObjectStore {
...
+        let metadata = CheckpointMetadata::decode(&metadata_bytes)?;
+
+        let checkpoint_data = match self.object_store.read(&path).await {
+            Ok(checkpoint) => {
+                // if the checksum exists, examine it.
+                if metadata.checksum != None {
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:433]:
```suggestion
                if let Some(checksum) = metadata.checksum {
```

File: src/mito2/src/manifest/storage.rs [hunk 9]
```diff
@@ -405,9 +417,42 @@ impl ManifestObjectStore {
...
+                    let calculated_checksum = get_checksum(&checkpoint);
+
+                    // verification process : succeed or not.
+                    if Some(format!("{}", calculated_checksum)) == metadata.checksum {
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:437]: You could use `ensure` to validate the checksum and return. What's more, I'd prefer `to_string()` here.
```suggestion
                    // verification process : succeed or not.
                    ensure!(checksum == calculated_checksum, VerifyChecksumSnafu { ... });
```

File: src/mito2/src/manifest/storage.rs [hunk 10]
```diff
@@ -419,42 +464,43 @@ impl ManifestObjectStore {
...
+                        let fall_back_path = gen_path(
+                            &self.path,
+                            &checkpoint_file(version),
+                            FALL_BACK_COMPRESS_TYPE,
+                        );
+                        debug!(
+                            "Failed to load checkpoint from path: {}, fall back to path: {}",
+                            path, fall_back_path
+                        );
+                        match self.object_store.read(&fall_back_path).await {
```
Comment (evenyag) [inline src/mito2/src/manifest/storage.rs:480]: We should also validate the checksum. Maybe we could add a method to get the checkpoint data first (compute checksum optionally), and then validate the checksum.
```rust
let calculate_checksum = metadata.checksum.is_some()
let Some((checkpoint_data, calculated_checksum)) = self.try_read_checkpoint_data(calculate_checksum)? else {
    return Ok(None);
};

// validate checksum...
```

File: src/mito2/src/manifest/tests/checkpoint.rs [hunk 11]
```diff
@@ -159,6 +160,53 @@ async fn manager_with_checkpoint_distance_1() {
...
+        .unwrap();
+    corrupted_bytes[0] ^= 1;
+
+    // Overwrite the latest checkpoint data
+    let _ = manager
+        .store()
+        .await
+        .write_last_checkpoint(9, &corrupted_bytes)
+        .await
+        .unwrap();
```
Comment (evenyag) [inline src/mito2/src/manifest/tests/checkpoint.rs:198]: Do we need this method? We could write to the object store using the last_checkpoint_path directly.

File: src/mito2/src/manifest/tests/checkpoint.rs [hunk 12]
```diff
@@ -159,6 +160,53 @@ async fn manager_with_checkpoint_distance_1() {
...
+
+    // Attempt to load the corrupted checkpoint
+    let load_corrupted_result = manager.store().await.load_last_checkpoint().await;
+
+    // Print load_corrupted_result for debugging
+    println!("Load Corrupted Result: {:?}", load_corrupted_result);
+
+    // Check if the result is an error and if it's of type VerifyChecksum
+    assert!(matches!(load_corrupted_result, Err(VerifyChecksum { .. })));
```
Comment (evenyag) [inline src/mito2/src/manifest/tests/checkpoint.rs:207]:
```suggestion
    // Check if the result is an error and if it's of type VerifyChecksum
    assert!(matches!(load_corrupted_result, Err(VerifyChecksum { .. })), "Load Corrupted Result: {:?}", load_corrupted_result);
```

Comment (CrystalAnalyst) [reply src/mito2/src/manifest/tests/checkpoint.rs:207]: > 

thanks for so many valuable suggestions for Improving !! I'll work on them next few days and I think I will learn a lot.

**Comment by WenyXu** [comment]
Hi, there are some conflicts. PTAL

**Pull request closed without merging**